src/cpp/shared/xml/utils_xml.h

```diff
@@ -1,16 +1,92 @@
 #ifndef UTILS_XML_H
 #define UTILS_XML_H
 
+#include "smacker.h"
 #include "tinyxml2/tinyxml2.h"
 
+#include <map>
+#include <set>
 #include <string>
+#include <vector>
 
 namespace UtilsXML {
   tinyxml2::XMLError QueryShortText(tinyxml2::XMLElement *el, short *dest);
   tinyxml2::XMLError QueryCharText(tinyxml2::XMLElement *el, char *dest);
   void QueryText(tinyxml2::XMLElement *el, char *dest);
   void QueryText(tinyxml2::XMLElement *el, std::string &dest);
-
+  template<typename T>
+  void PushBack(tinyxml2::XMLDocument *doc, tinyxml2::XMLNode *dest, const char* name, const T &val) {
+    tinyxml2::XMLElement *elem = doc->NewElement(name);
+    elem->SetText(val);
+    dest->InsertEndChild(elem);
+  }
+  template <typename T>
+  void WriteArray(tinyxml2::XMLDocument *doc, tinyxml2::XMLNode *dest, const char* name, const T &src) {
+    for(int i = 0; i < ELEMENTS_IN(src); i++) {
```

Comment:
Are you sure that ELEMENTS_IN will work here?

Reply:
It requires that the array's size is statically known.

Reply:
Compiler makes functions from this template for each array type like char[8], short[11] e.t.c.

Reply:
template<typename T, size_t N>  void foo(const T (&t)[N]) { ... }

Declare it like this. Then you can't pass in something that's not an array here.

Reply:
Fixed here and also in WriteCampaignDDArray